### PR TITLE
Add the ability to run tasks in given apps for umbrella applications

### PR DESCRIPTION
**Pull request by jmbejar**
This PR implements an idea that was proposed in the discussion about #10984.

The idea is to add the `--app` option to the `do` task, so we have more control to indicate where tasks are executed (which makes sense in the context of umbrella projects).

#10984 is solved in an indirect way because this solution will emit warnings when the given apps do not exist, something that is impossible in the context of the `cmd` task (the other task that accepts `--app` so far). So, we would need to teach that it is better to run tasks (even the `cmd task`) via `mix do` if we want to indicate in which apps tasks should run.

## Examples

```bash
# Basic example in the context of an umbrella application with two apps: my_app and second_app

mix do --app my_app --app second_app test, cmd echo hello

==> my_app
...
1 doctest, 1 test, 0 failures
==> second_app
...
1 doctest, 1 test, 0 failures
==> my_app
hello
==> second_app
hello
```

```bash
# Running the tasks in one specific app

mix do --app my_app test, cmd echo hello

==> my_app
...
1 doctest, 1 test, 0 failures
==> my_app
hello
```

```bash
# Passing one application name with a typo

mix do --app my_app --app zecond_app test, cmd echo hello

warning: could not find umbrella app :zecond_app
==> my_app
...
1 doctest, 1 test, 0 failures
==> my_app
hello
```

```bash
# Invoking `mix do` with apps options, but the project is not an umbrella application

mix do --app my_app help, cmd echo hello

warning: running "help" at root level because this is not an umbrella project
<`mix help` output here>
warning: running "cmd" at root level because this is not an umbrella project
hello
```

## Open questions

* Should we deprecate the `--app` option in `mix cmd`? I think it is not completely necessary, but it feels like a bit of duplication in the API (since it would be possible to run the `cmd` task with `mix do` if we need to specify apps).

### CODE REVIEW

File: lib/mix/lib/mix/tasks/do.ex [hunk 1]
```diff
@@ -28,7 +28,20 @@ defmodule Mix.Tasks.Do do
   @impl true
   def run(args) do
     Mix.Task.reenable("do")
-    Enum.each(gather_commands(args), fn [task | args] -> Mix.Task.run(task, args) end)
+
+    {apps, args} = extract_apps_from_args(args)
```
Comment (jmbejar) [inline lib/mix/lib/mix/tasks/do.ex:32]: We need to write about this new option in the module doc.

File: lib/mix/test/mix/tasks/do_test.exs [hunk 2]
```diff
@@ -24,4 +24,127 @@ defmodule Mix.Tasks.DoTest do
...
+                           "warning: running \"cmd\" at root level because this is not an umbrella project"
+                         ]}
+
+        assert_received {:mix_shell, :info,
+                         [
+                           "warning: running \"p\" at root level because this is not an umbrella project"
+                         ]}
+      end)
+    end)
+  end
```
Comment (jmbejar) [inline lib/mix/test/mix/tasks/do_test.exs:149]: I need to confirm we have all cases (code branches) covered with these test cases.

File: lib/mix/lib/mix/task.ex [hunk 3]
```diff
@@ -445,32 +499,60 @@ defmodule Mix.Task do
...
 
-  defp run_alias([h | t], alias_args, proj, original_task, _res) when is_function(h, 1) do
-    res = h.(join_args([], alias_args, t))
-    run_alias(t, alias_args, proj, original_task, res)
+  defp run_alias([h | t], alias_args, proj, original_task, options, _res)
+       when is_function(h, 1) do
+    res =
+      case options do
+        [apps: apps] ->
+          if Mix.Project.umbrella?() do
```
Comment (jmbejar) [inline lib/mix/lib/mix/task.ex:527]: I think the code here is hard to read and follow. I will work on it a little bit.

File: lib/mix/lib/mix/task.ex [hunk 4]
```diff
@@ -347,29 +358,46 @@ defmodule Mix.Task do
...
+
+  @doc """
+  Similar to run/2 but it runs the task in the specified list of children apps,
+  in the context of umbrella proyects.
+
+  However, if the task is not recursive (whose purpose is to be run in children
+  applications) it runs at the project root level (a warning message is shown
+  in this case).
+  """
+  @spec run_in_apps(task_name, [atom], [any]) :: any
```
Comment (jmbejar) [inline lib/mix/lib/mix/task.ex:373]: Do we need to add `@doc since: ..`? Which version?

File: lib/mix/lib/mix/task.ex [hunk 5]
```diff
@@ -347,29 +358,46 @@ defmodule Mix.Task do
...
+  def run_in_apps(task, apps, args \\ []) do
+    do_run(task, args, apps: apps)
+  end
 
-  def run(task, args) when is_atom(task) do
-    run(Atom.to_string(task), args)
+  defp do_run(task, args, options \\ [])
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:378]: We typically don't use default arguments for private functions. Instead we change all callers. Personally speaking, I would remove the `options` and call it `in_apps`. it is either `nil` or a `list`. This should simplify a bit the code below and the amount of state kept.

File: lib/mix/lib/mix/task.ex [hunk 6]
```diff
@@ -380,12 +408,38 @@ defmodule Mix.Task do
...
 
+      in_apps && !Mix.Project.umbrella?() ->
+        Mix.shell().info([
+          :yellow,
+          "warning: running #{inspect(task)} at root level because this is not an umbrella project"
+        ])
+
+        run(task, args)
+
+      in_apps && not recursive ->
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:435]: I think we should not care about this. Sometimes if the task is recursive is an implementation detail. We should just not emit any warning. :)

File: lib/mix/lib/mix/task.ex [hunk 7]
```diff
@@ -445,32 +499,60 @@ defmodule Mix.Task do
...
-        res = run(task, join_args(args, alias_args, t))
-        run_alias(t, alias_args, proj, original_task, res)
+        res =
+          case options do
+            [apps: apps] ->
+              run_in_apps(task, apps, join_args(args, alias_args, t))
+
+            _ ->
+              run(task, join_args(args, alias_args, t))
+          end
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:516]: If you make `options` be a single `in_apps` variable that is either `nil` or a list of apps, you can rewrite this to:

    do_run(task, join_args(args, alias_args, t), in_apps)

File: lib/mix/lib/mix/task.ex (hunk 3 above)
Comment (josevalim) [reply lib/mix/lib/mix/task.ex:527]: I think if you replace options by `in_apps` you can have a nice `cond`:

```elixir
cond do
  in_apps && Mix.Project.umbrella?() ->

  in_apps ->

  true ->

end
```

File: lib/mix/lib/mix/task.ex [hunk 8]
```diff
@@ -522,12 +604,26 @@ defmodule Mix.Task do
...
 
   defp recur(fun) do
+    run_in_children_projects(fun, Mix.Dep.Umbrella.cached())
+  end
+
+  defp recur_in_apps(fun, apps) do
+    selected_children =
+      Mix.Dep.Umbrella.cached()
+      |> Enum.filter(fn %Mix.Dep{app: app} ->
+        app in apps
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:614]: Should we warn if one of the apps do not exist?

File: lib/mix/lib/mix/task.ex [hunk 9]
```diff
@@ -326,6 +326,17 @@ defmodule Mix.Task do
     end
   end
 
+  @doc false
+  def show_forgotten_apps_warning(apps) do
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:330]: I would move this warning to `run_in_apps`. I dropped a comment below. :)

File: lib/mix/lib/mix/task.ex (hunk 4 above)
Comment (josevalim) [reply lib/mix/lib/mix/task.ex:373]: Yes. 1.14.0!

**Comment by josevalim** [comment]
This is looking great! I have dropped some comments that can help clean up the code a bit. :)

File: lib/mix/lib/mix/task.ex (hunk 9 above)
Comment (jmbejar) [reply lib/mix/lib/mix/task.ex:330]: This was my first attempt, but I noticed two situations where the warning is displayed multiple times:

* The warning will be displayed for each different task because `run_in_apps` is invoked from `Task.Mix.Do` for each passed task.
* There is a recursive call from the aliases logic (https://github.com/elixir-lang/elixir/pull/11593/files#diff-4b0f94e7f3d2186a8e00b12c6b7c9bda8a7d86e8bb24f8013cf16b405bbdee8aR512), so the warning will be displayed twice in this case, as well.

So, I wanted to make sure the warning happens only once. Note the warning is about the no-exiting apps, so it has to be displayed/controlled at "root level" (in this case, in the context of the initial call from `Task.Mix.Do`).

Does it make sense?

Comment (josevalim) [reply lib/mix/lib/mix/task.ex:330]: Yes, it does. So maybe I would keep it in the `do` task then. I think you can use `Mix.Project.apps_paths` to get the current app names.

File: lib/mix/lib/mix/task.ex (hunk 3 above)
Comment (jmbejar) [reply lib/mix/lib/mix/task.ex]: Thanks for the suggestion. Everything is simpler now with the removal of the options arg 👍

File: lib/mix/lib/mix/task.ex (hunk 8 above)
Comment (jmbejar) [reply lib/mix/lib/mix/task.ex:579]: Checks live now in `Mix.Task.Do`

File: lib/mix/lib/mix/task.ex [hunk 10]
```diff
@@ -347,29 +347,47 @@ defmodule Mix.Task do
...
+  applications) it runs at the project root level (a warning message is shown
+  in this case).
+  """
+  @doc since: "1.14.0"
+  @spec run_in_apps(task_name, [atom], [any]) :: any
+  def run_in_apps(task, apps, args \\ []) do
+    do_run(task, args, apps)
+  end
+
+  defp do_run(task, args, apps \\ nil)
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:368]: Small nitpick: we don't use default arguments for private functions. Instead change the callers to pass the desired argument explicitly. :)

File: lib/mix/lib/mix/task.ex [hunk 11]
```diff
@@ -445,32 +474,51 @@ defmodule Mix.Task do
...
+        apps && Mix.Project.umbrella?() ->
+          Mix.ProjectStack.recur(fn ->
+            recur(fn _ -> h.(join_args([], alias_args, t)) end, apps)
+          end)
+
+        apps ->
+          Mix.shell().info([
+            :yellow,
+            "warning: running #{inspect(original_task)} at root level because this is not an umbrella project"
+          ])
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:502]: I was thinking we should refactor this warning into a private function... but then I am wondering, maybe this should be an error?

File: lib/mix/lib/mix/task.ex [hunk 12]
```diff
@@ -521,13 +569,26 @@ defmodule Mix.Task do
     :ok
   end
 
-  defp recur(fun) do
+  defp recur(fun, apps \\ nil)
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:572]: Same note here, don't use default args for private functions. :)

File: lib/mix/lib/mix/task.ex [hunk 13]
```diff
@@ -521,13 +569,26 @@ defmodule Mix.Task do
...
+  defp recur(fun, nil), do: run_in_children_projects(fun, Mix.Dep.Umbrella.cached())
+
+  defp recur(fun, apps) do
+    selected_children =
+      Mix.Dep.Umbrella.cached()
+      |> Enum.filter(fn %Mix.Dep{app: app} ->
+        app in apps
+      end)
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:580]:
```suggestion
      |> Enum.filter(fn %Mix.Dep{app: app} -> app in apps end)
```

File: lib/mix/lib/mix/tasks/do.ex [hunk 14]
```diff
@@ -15,20 +15,63 @@ defmodule Mix.Tasks.Do do
...
+    if Mix.Project.umbrella?() do
+      forgotten_apps = apps -- Enum.map(Mix.Dep.Umbrella.cached(), & &1.app)
+
+      for app <- forgotten_apps do
+        Mix.shell().info([:yellow, "warning: could not find application #{inspect(app)}"])
+      end
+    end
+  end
+
+  defp extract_apps_from_args(args) do
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/do.ex:66]: I think we want to parse the args explicitly. In particular, if we have a task like this:

    mix do --app foo my_task --app bar

The `--app foo` should be captured, but not `--app bar`. Recursion should be enough to do it.

File: lib/mix/lib/mix/tasks/do.ex [hunk 15]
```diff
@@ -15,20 +15,63 @@ defmodule Mix.Tasks.Do do
...
 
       mix do compile, some_other_command, compile
 
   When `compile` is executed again, Mix will notice the task
   has already ran, and skip it.
+
+  ## Command line options
+
+    * `--app` - limit running the tasks to the given app. This
+    option may be given multiple times.
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/do.ex:35]:
```suggestion
      option may be given multiple times and must come before any of the tasks.
```

Comment (josevalim) [inline lib/mix/lib/mix/tasks/do.ex:35]:
```suggestion
    * `--app` - limit running the tasks to the given app. This option may
      be given multiple times and must come before any of the tasks.
```

File: lib/mix/lib/mix/task.ex (hunk 11 above)
Comment (josevalim) [reply lib/mix/lib/mix/task.ex:502]: The advantage of making it an error is that we can check it once at the top of `run_in_apps` and just raise. In here we can assume if there are `apps`, it is an umbrella.

File: lib/mix/lib/mix/tasks/do.ex (hunk 14 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/do.ex:66]: You can test this with something like: `mix do cmd echo --app bar` and that should not run inside an app.

**Comment by josevalim** [comment]
Thank you @jmbejar and sorry for the delay! I have done a last batch of reviews, mostly minor details pending. :)

File: lib/mix/test/mix/tasks/do_test.exs [hunk 16]
```diff
@@ -24,4 +24,115 @@ defmodule Mix.Tasks.DoTest do
...
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+
+        assert_received {:mix_shell, :info, ["==> foo"]}
+        assert_received {:mix_shell, :run, ["hello" <> ^nl]}
+        assert_received {:mix_shell, :info, ["[\"Foo\"]"]}
+      end)
+    end)
+  end
+
+  test "runs non-recursive aliases at project root level" do
```
Comment (jmbejar) [inline lib/mix/test/mix/tasks/do_test.exs:108]: @josevalim Do you still think this edge case is well resolved? I initially emitted a warning for this kind of tasks (because the -app argument is basically ignored here) but we reverted because of this https://github.com/elixir-lang/elixir/pull/11593#pullrequestreview-862038204

Since we are now raising an error when --app is given in no-umbrella projects, I was thinking again in this situation and I don't like the fact we are still running the task (at the root level) and ignoring the --app option, without any kind of warning/error.

Maybe you're still OK, I'm just mentioning it just in case you want to double check :)

File: lib/mix/lib/mix/tasks/do.ex (hunk 14 above)
Comment (jmbejar) [reply lib/mix/lib/mix/tasks/do.ex:66]: > I think we want to parse the args explicitly. In particular, if we have a task like this:
> 
> ```
> mix do --app foo my_task --app bar
> ```
> 
> The `--app foo` should be captured, but not `--app bar`. Recursion should be enough to do it.

I think we are already doing exactly that. In fact, I copied the approach from the `cmd` task (where we already have the --app option). It relies on `OptionParser.parse_head!`.

I tested it directly in iex, and here is the result (the first element of the result tuple are the captures):

```ex
iex(16)> OptionParser.parse_head!(["--app", "foo", "my_task", "--app", "bar"], switches: [app: :keep])
{[app: "foo"], ["my_task", "--app", "bar"]}

iex(17)> OptionParser.parse_head!(["cmd", "echo", "--app", "bar"], switches: [app: :keep])
{[], ["cmd", "echo", "--app", "bar"]}
```

Comment (jmbejar) [reply lib/mix/lib/mix/tasks/do.ex:66]: > You can test this with something like: `mix do cmd echo --app bar` and that should not run inside an app.

See the results running in a test project. It is in line with my previous comment (the --app bar is not captured by the `do` task).

```
➜  mix do cmd echo --app bar
==> second_app
--app bar
==> my_app
--app bar
```

Note:  The `that should not run inside an app` is not accurate. The task is run **in every application**, as it currently works in master (where the --app does not exist at all).

**Comment by jmbejar** [comment]
> Thank you @jmbejar and sorry for the delay! I have done a last batch of reviews, mostly minor details pending. :)

No problem! I tried to make as much progress as I could, but some new questions/comments came up.

File: lib/mix/lib/mix/task.ex [hunk 17]
```diff
@@ -445,32 +470,43 @@ defmodule Mix.Task do
...
+    res =
+      cond do
+        apps ->
+          Mix.ProjectStack.recur(fn ->
+            recur(fn _ -> h.(join_args([], alias_args, t)) end, apps)
+          end)
+
+        true ->
+          h.(join_args([], alias_args, t))
+      end
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex:496]:
```suggestion
      if apps do
        Mix.ProjectStack.recur(fn ->
          recur(fn _ -> h.(join_args([], alias_args, t)) end, apps)
        end)
      else
        h.(join_args([], alias_args, t))
      end
```

File: lib/mix/test/mix/tasks/do_test.exs (hunk 16 above)
Comment (josevalim) [reply lib/mix/test/mix/tasks/do_test.exs:108]: For now I think this is the expected behaviour. If a task is recursive or not is kind of an implementation detail. We will see what comes out of this in practice. :)

File: lib/mix/lib/mix/tasks/do.ex (hunk 14 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/do.ex:66]: Ah, I see. This is a smart trick!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
